File: Easy/rupsaTri/rupsaTri.java
//run as java -ea rupsaTri
import java.util.Scanner;

public class rupsaTri{
	public static void main(String[] args){
		int T,N;
		Scanner scan = new Scanner(System.in);

		boolean[] P = new boolean[5000001];
		boolean[] Q = new boolean[5000001];

		for(int i=2;i<2237;i++){
			if(P[i]==true){
				continue;
			}

			for(int j=2;i*j<5000001;j++){
				P[i*j] = true;
			}
		}

		for(int i=2;i<5000001;i++){
			if(P[i]==true){
				continue;
			}

			if((i-1)%4==0){
				for(int j=1;i*j<5000001;j++){
					Q[i*j] = true;
				}
			}
		}

		T = scan.nextInt();
		assert(T>=1 && T<=1000000) : "T must be 1 <= T <= 10^6";

		while(T>0){
			N = scan.nextInt();
			assert(N>=1 && N<=5000000) : "N must be 1 <= N <= 5*10^6";

			if(Q[N]==true){
				System.out.println("YES");
			}else{
				System.out.println("NO");
			}

			T--;
		}
	}
}

File: Easy/rupsaTri/rupsaTri.c
#include <stdio.h>
#include <assert.h>

int check[5000005] = {0};

int main(void){
	int T,N;

	for(int i=1;i<=2237;i++){
		for(int j=i;j<=2237;j++){
			if((i*i)+(j*j)>=5000005){
				break;
			}

			if(i!=j){
				check[(i*i)+(j*j)] = 1;
			}
		}
	}

	check[2] = 0;

	for(int i=2500001;i>1;i--){
		if(check[i]!=0){
			for(int j=2;i*j<5000001;j++){
				check[i*j] = 1;
			}
		}
	}

	scanf("%d",&T);
	assert(T>=1 && T<=1000000);

	while(T>0){
		scanf("%d",&N);
		assert(N>=1 && N<=5000000);

		if(check[N]!=0){
			printf("YES\n");
		}else{
			printf("NO\n");
		}

		T--;
	}
	
	return 0;
}